### DISCUSSION TITLE
Why Doesn't My RecipeType work?

### DISCUSSION THREAD

**Discussion by mangodev01** [FabricMC/fabric-api#2484] [Mod Dev Support]
i wanted to make a RecipeType for my diamond crafting table in my mod, for some reason it isn't working, can anyone help please?,

the source code is [here](https://github.com/Gamer069/illia_fabric_1_19_mod)

**Comment by warjort** [comment]
You already have the code for that here: 
https://github.com/Gamer069/illia_fabric_1_19_mod/blob/033026eb14b2f5f30dd1dab8a2f6b7c667631e0e/src/main/java/net/illia/illiafabricmod1_19/screen/DiamondCraftingTableScreenHandler.java#L102

But it has 2 problems (I guess you just copied it from the vanilla crafting table?):
* It uses  RecipeType.CRAFTING, when you want it to use your RecipeType
* Nothing actually calls that method. :-)

If you follow what the vanilla crafting table does, you could do something like:
````
@Override
public void onContentChanged(Inventory inventory) {
    updateResult(this, this.player.world, this.player, this.input, this.result);
}
````

First check that makes vanilla crafting recipes work with your block.

But you should check with the vanilla crafting table code. It also has a check that the block isn't destroyed.
e.g. a creeper blew up your block while you were crafting. :-)

This seems to be entirely missing in your code.
See CraftingScreenHandler.canUse()
You don't have a ScreenHandlerContext in DiamondCraftingTableScreenHandler?

Once you get that working, you want to change the RecipeType to this one:
https://github.com/Gamer069/illia_fabric_1_19_mod/blob/033026eb14b2f5f30dd1dab8a2f6b7c667631e0e/src/main/java/net/illia/illiafabricmod1_19/recipe/DiamondCraftingTableRecipe.java#L79

**Reply by mangodev01** [reply]
so, then do i even need a recipe type?

**Reply by warjort** [reply]
If you just want vanilla recipes, you don't need your own recipe type.

**Reply by mangodev01** [reply]
thanks, do i just add 
```
	public void onContentChanged(Inventory inventory) {
		this.context.run(((world, pos) -> {
			updateResult(this, world, this.player, this.input, this.result);
		}));
	}
```
,  make a context (ScreenHandlerContext) to fix the issue?

**Reply by warjort** [reply]
That is what the vanilla table does, yes.

**Reply by mangodev01** [reply]
thanks

**Reply by mangodev01** [reply]
but now i get this error when i replace
```
RecipeType.CRAFTING
```
with
```
Type.INSTANCE
```

```
'getFirstMatch(net.minecraft.recipe.RecipeType<T>, C, net.minecraft.world.World)' in 'net.minecraft.recipe.RecipeManager' cannot be applied to '(net.illia.illiafabricmod1_19.recipe.DiamondCraftingTableRecipe.Type, net.minecraft.inventory.CraftingInventory, net.minecraft.world.World)'
```

**Reply by warjort** [reply]
This is a compiler error because your types are wrong. This a java question again.
You don't show the code to give any context. Fortunately, I can deduce what is wrong :-) 
But in future you need to update your repository so we can see what you have done.

1) Your recipe should extend `Recipe<CraftingInventory>` here:
https://github.com/Gamer069/illia_fabric_1_19_mod/blob/033026eb14b2f5f30dd1dab8a2f6b7c667631e0e/src/main/java/net/illia/illiafabricmod1_19/recipe/DiamondCraftingTableRecipe.java#L14
and modify the class accordingly.
This is the type of "craftingInventory" you are passing to that method.

2) You can't just change one line of that method. That is just the main operational change.
There are references to CraftingRecipe you need to change to DiamondCraftingTableRecipe as a consequence of that change.

**Reply by mangodev01** [reply]
thanks

**Reply by mangodev01** [reply]
now the game launches, but now for some reason my block entity is broken, when i place it then the game crashes
